content/blog/2022-03-whats-new-in-axum-0-5.md

```diff
@@ -0,0 +1,298 @@
+        SetHeader("x-bar", "another custom header"),
+        Json(json!({})),
+    )
+}
+```
+
+`IntoResponseParts` is also implemented for [`Extension`] making it easy to set
+response extensions, for example to share state with middleware:
+
+```rust
+use axum::{Extension, Json, response::IntoResponse};
+use serde_json::json;
+
+async fn json_extensions() -> impl IntoResponse {
+    (
+        Extension(some_value),
+        Extension(some_other_value),
+        Json(json!({})),
+    )
+}
+```
+
+See [`axum::response`] for more details.
+
+## Cookies
+
+Building on top of `IntoResponseParts` [axum-extra] has a new [`CookieJar`]
+extractor:
+
+```rust
+use axum_extra::extract::cookie::{CookieJar, Cookie};
+use axum::response::IntoResponseParts;
+
+async fn handler(jar: CookieJar) -> impl IntoResponse {
+    if let Some(cookie_value) = jar.get("some-cookie") {
+        tracing::info!(?cookie_value);
+    }
+
+    let updated_jar = jar
+        .add(Cookie::new("session_id", "value"))
+        .remove(Cookie::named("some-cookie"));
+
+    (updated_jar, "response body...")
+}
+```
+
+It also comes in a [`SignedCookieJar`] variant that will sign cookies with a key so
+you're sure someone hasn't tampered with them.
+
+`IntoResponseParts` makes this possible without requiring any middleware.
+
+See [`axum_extra::extract::cookie`] for more details.
+
+## `HeaderMap` extractor
+
+You've always been able to use [`HeaderMap`] as an extractor to access headers
+from the request. But what you might not realise is that this would implicitly
+consume the headers such that other extractors wouldn't be able to access them.
+
+For example this is subtly broken:
+
+```rust
+use axum::{http::HeaderMap, extract::Form};
+
+async fn handler(
+    headers: HeaderMap,
+    form: Form<Payload>,
+) {
+    // ...
+}
+```
+
+Since we run the `HeaderMap` first, `Form` would be unable to access them and
+fail with a `500 Internal Server Error`. This was quite surprising and caused
+headaches for some users.
+
+However in axum 0.5 this problem goes away and it just works!
+
+## More flexible `Router::merge`
+
+[`Router::merge`] can be used to merge two routers into one. In axum 0.5 its
+gotten slightly more flexible since it now accepts any `impl Into<Router>`. This
+allows you to have custom ways of constructing `Router`s and have them work
+seamlessly with axum.
+
+One could imagine a way to compose REST and gRPC like so:
+
+```rust
+let rest_routes = Router::new().route(...);
+
+// with `impl From<GrpcService> for Router`
+let grpc_service = GrpcService::new(GrpcServiceImpl::new());
+
+let app = Router::new()
+    .merge(rest_routes)
+    .merge(grpc_service);
+```
+
+## Honorable mentions
+
+The following features weren't new in 0.5 but shipped recently and are worthy of
+a shout out.
+
+### `middleware::from_fn`
+
+axum uses the [`tower::Service`] trait for middleware however it can be little
+daunting to implement, mainly due to the lack of async traits in Rust.
```

Comment:
```suggestion
`axum` uses the [`tower::Service`] trait for middleware. However, it can be little
daunting to implement, mainly due to the lack of async traits in Rust.
```